content/en/blog/_posts/2021-08-30-volume-populators-alpha.md

```diff
@@ -0,0 +1,207 @@
+---
+title: Volume Populators Redesign
+date: 2021-08-30
+---
+
+**Authors:**
+Ben Swartzlander (NetApp)
+
+Kubernetes v1.22 introduces a redesigned approach for volume populators. Originally implemented
+in v1.18, the API suffered from backwards compatibility issues and v1.22 includes a new API
+field called `dataSourceRef` that fixes these problems.
+
+## Data sources
+
+The PersistentVolumeClaim object got a new field called `dataSource` when snapshot and cloning
+support was added to Kubernetes. This field allowed new PVCs to point to other PVCs or
+VolumeSnapshots in the same namespace, and modified the normal provisioning process so that
+instead of yielding an empty volume, the new PVC contained the same data as either the cloned
+PVC or the cloned VolumeSnapshot.
+
+Volume populators embrace the same design idea, but extend it to any type of object, as long
+as there exists a CRD to define the data source and a populator controller to implement the
+logic. Initially the `dataSource` field was directly extended to allow arbitrary objects, if
+the `AnyVolumeDataSource` feature gate was enabled on the cluster. That change unfortunately
+caused backwards compatibility problems, and so the new `dataSourceRef` field was born.
+
+In v1.22 if the `AnyVolumeDataSource` feature gate is enabled, the `dataSourceRef` field is
+added, which behaves similarly to the `dataSource` field except that it allows arbitrary
+objects to be specified. The API server ensures that the two fields always have the same
+contents, and neither of them are mutable. The differences is that at creation time
+`dataSource` allows only PVCs or VolumeSnapshots, and ignores all other values, while
+`dataSourceRef` allows most types of objects, and in the few cases it doesn't allow an
+object (core objects other than PVCs) a validation error occurs.
+
+This API redesign will eventually result in a preference for the `dataSourceRef` field for
+all use cases, but in the immediate timeframe it is used by volume populators.
```

Comment:
I'm not sure what you are trying to say here.  You meant dataSourceRef is used only for create volume from AnyVolumeDataSource now, but eventually dataSourceRef will be the preferred one for create PVC from VolumeSnapshot, cloning, and volume populators?

Reply:
What I'm trying to say is that for clusters where the feature gate is enabled, dataSourceRef is the preferred API. In the future (wen the feature is GA) that will be all clusters. However that's the vast minority of clusters while the feature is Alpha.